dashboard/app/views/ops_mailer/_exit_survey_url.html.haml

```diff
@@ -1,52 +1,63 @@
-- case workshop.program_type.to_i
-  - when 1
-    %p
-      We appreciate any feedback you have. Please take a moment to complete this
-      - case workshop.phase
-        - when 2
+%p
+  We appreciate any feedback you have. Please take a moment to complete this
+  - case workshop.program_type_short_name
+    - when 'CS in Science'
+      - case workshop.phase_short_name
+        - when 'Phase 2'
           %a{href:"https://docs.google.com/a/code.org/forms/d/1QG9eCbKJD26UNvTC0C9ZZyrp63WjzQSK5gQPP4lsZ2c/viewform"}
             #{link_text}
-          about your Code.org workshop experience.
-        - when 4
+        - when 'Phase 3B'
+          %a{href:"https://docs.google.com/a/code.org/forms/d/1V4sm9cPvABp4ipo4MIMUq9cf-aa-sFWafI5qjrlSV8M/viewform?c=0&w=1"}
+            #{link_text}
+        - when 'Phase 4'
           %a{href:"https://docs.google.com/a/code.org/forms/d/106j69amXoeZkBEXfNszLUpIx3FfyVcOEdo5Ce6oKMA4/viewform?c=0&w=1&usp=send_form"}
             #{link_text}
-          about your Code.org workshop experience.
-  - when 2
-    %p
-      We appreciate any feedback you have. Please take a moment to complete this
-      - case workshop.phase
-        - when 2
+    - when 'CS in Algebra'
+      - case workshop.phase_short_name
+        - when 'Phase 2'
           %a{href:"https://docs.google.com/a/code.org/forms/d/1kEkyOk-Y4MsWVVjmkaw-UDfCE4Wl2Ol3MTrCw4IW8vw/viewform"}
             #{link_text}
-        - when 4
+        - when 'Phase 3B'
+          %a{href:"https://docs.google.com/a/code.org/forms/d/12koHnpyF2azrNNsEgvy66_dqXXmWeRmEJ9ICJA-gT-M/viewform?c=0&w=1"}
+            #{link_text}
+        - when 'Phase 4'
           %a{href:"https://docs.google.com/a/code.org/forms/d/1oBYFuv1xlInKD7asJ7gR4GA1xbsdKDjG1PI4tvgdwWc/viewform?c=0&w=1"}
             #{link_text}
-          about your Code.org workshop experience.
-  - when 3
-    %p
-      We appreciate any feedback you have. Please take a moment to complete this
-      - case workshop.phase
-        - when 2
+    - when 'ECS'
+      - case workshop.phase_short_name
+        - when 'Phase 2'
           %a{href:"https://docs.google.com/a/code.org/forms/d/12-HzFgPn6vP-D5P1XCkYxg_Dv3_DWM0B-xh1VDBuhNo/viewform"}
```

Comment:
Sure -- so there are bunch of things that are saying "refactor" to me:
* visual patterns looking at code == repetition
* lots of logic in a view == violates MVC (you also noticed yesterday that writing a test for this was awkward because you had to look at the contents of the email)
* conditionals with lots of branching (this is essentially a conditional with 17 branches)

So, first thing I would do is move this stuff out of the view and in to the model. So it would look something like:

app/views/ops_mailer/_exit_survey_url.html.haml:
````
 %a{href: @workshop.exit_survey_url} =link_text
````

(this partial is now only 4 lines and it appears that `link_text` is always the same, so I think it would prob make sense to refactor this to eliminate this partial)

app/models/workshop.rb:
````
class Workshop
[...]
  def exit_survey_url
    # copy the case statement in here, essentially
    case program_type_short_name
      when 'CS in Science'
        case phase_short_name
          when 'Phase 2'
         "https://docs.google.com/a/code.org/forms/d/1QG9eCbKJD26UNvTC0C9ZZyrp63WjzQSK5gQPP4lsZ2c/viewform"
          when 'Phase 3B'
            "https://docs.google.com/a/code.org/forms/d/1V4sm9cPvABp4ipo4MIMUq9cf-aa-sFWafI5qjrlSV8M/viewform?c=0&w=1"
[...]
````
conveniently you can now test this method without messing around with mail sending too (it might be worth adding a simple test at this point before we go refactoring crazy)...

So now the view logic problem is gone... next is the mega case statement. I would start by just putting it in a new hash, which could look like:

````
````
class Workshop
[...]
  EXIT_SURVEY_URLS = {
    'CS in Science' => {
      'Phase 2' => 'https://docs.google.com/a/code.org/forms/d/1QG9eCbKJD26UNvTC0C9ZZyrp63WjzQSK5gQPP4lsZ2c/viewform',
     'Phase 3B' => 
'https://docs.google.com/a/code.org/forms/d/1V4sm9cPvABp4ipo4MIMUq9cf-aa-sFWafI5qjrlSV8M/viewform?c=0&w=1',
[ ...]
    }
  }
  def exit_survey_url
     if EXIT_SURVEY_URL[program_type_short_name]
       EXIT_SURVEY_URL[phase_short_name]
     end
  end
    # copy the case statement in here, essentially

````

So next I would maybe look at whether this has anything in common with the other hashes in WorkshopConstants, and maybe move it in there if it looks like it makes sense for all the workshop constants to be together, and maybe see if any of the hashes can be merged and/or generated from one source (ie. the interface to WorkshopConstants stays the same, but the three hashes are initialized from one thing) or if it looks better to have them separate.

After that, what would look weird to me:
* there are still lots of magic strings and integers going on for program type and phase -- it would be really easy to make a typo, eg put a survey under "Phase 3 B" instead of "Phase 3B" -- maybe turn them into constants? Or maybe not if it's all encapsulated in WorkshopConstants.
* All the urls are almost the same except for the id and url params -- do the url params mean anything (and should they actually all be the same), and might it make sense to store only the id per program type/phase, and generate the rest of the url? Maybe not if we don't want to get into the 'meaning' of this url here.
* something up there might need a `if workshop.exit_survey_url` around it -- current behavior for a missing exit survey url appears to be to still print the text around the exit survey url but not include the link.